Source/Core/DolphinQt2/HotkeyScheduler.cpp

```diff
@@ -0,0 +1,350 @@
+
+#include "AudioCommon/AudioCommon.h"
+#include "Common/Thread.h"
+#include "Core/Core.h"
+#include "Core/HotkeyManager.h"
+#include "Core/IOS/IOS.h"
+#include "Core/IOS/USB/Bluetooth/BTBase.h"
+#include "Core/State.h"
+#include "DolphinQt2/MainWindow.h"
+#include "DolphinQt2/Settings.h"
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+#include "VideoCommon/VertexShaderManager.h"
+#include "VideoCommon/VideoConfig.h"
+
+HotkeyScheduler::HotkeyScheduler(MainWindow* parent) : m_parent(parent)
+{
+  HotkeyManagerEmu::Initialize();
+  HotkeyManagerEmu::LoadConfig();
+  HotkeyManagerEmu::Enable(true);
+}
+
+void HotkeyScheduler::Start()
+{
+  m_stop_requested = false;
+  m_thread = std::thread(&HotkeyScheduler::Run, this);
+  m_thread.detach();
+}
+
+void HotkeyScheduler::Stop()
+{
+  m_stop_requested = true;
+}
+
+static bool IsHotkey(int id, bool held = false)
+{
+  return HotkeyManagerEmu::IsPressed(id, held);
+}
+
+static void HandleFrameskipHotkeys()
+{
+  constexpr int MAX_FRAME_SKIP_DELAY = 60;
+  constexpr int FRAME_STEP_DELAY = 30;
+
+  static int frame_step_count = 0;
+  static int frame_step_delay = 1;
+  static int frame_step_delay_count = 0;
+  static bool frame_step_hold = false;
+
+  if (IsHotkey(HK_FRAME_ADVANCE_INCREASE_SPEED))
+  {
+    frame_step_delay = std::min(frame_step_delay + 1, MAX_FRAME_SKIP_DELAY);
+    return;
+  }
+
+  if (IsHotkey(HK_FRAME_ADVANCE_DECREASE_SPEED))
+  {
+    frame_step_delay = std::max(frame_step_delay - 1, 0);
+    return;
+  }
+
+  if (IsHotkey(HK_FRAME_ADVANCE_RESET_SPEED))
+  {
+    frame_step_delay = 1;
+    return;
+  }
+
+  if (IsHotkey(HK_FRAME_ADVANCE, true))
+  {
+    if (frame_step_delay_count < frame_step_delay && frame_step_hold)
+      frame_step_delay_count++;
+
+    // TODO GUI Update (Depends on an unimplemented feature)
+    // if ((frame_step_count == 0 || frame_step_count == FRAME_STEP_DELAY) && !frame_step_hold)
+
+    if (frame_step_count < FRAME_STEP_DELAY)
+    {
+      ++frame_step_count;
+      if (frame_step_hold)
+        frame_step_hold = false;
+    }
+
+    if (frame_step_count == FRAME_STEP_DELAY && frame_step_hold &&
+        frame_step_delay_count >= frame_step_delay)
+    {
+      frame_step_hold = false;
+      frame_step_delay_count = 0;
+    }
+
+    return;
+  }
+
+  if (frame_step_count > 0)
+  {
+    // Reset frame advance
+    frame_step_count = 0;
+    frame_step_hold = false;
+    frame_step_delay_count = 0;
+  }
+}
+
+void HotkeyScheduler::Run()
+{
+  while (!m_stop_requested)
+  {
+    if (!HotkeyManagerEmu::IsEnabled())
+      continue;
+
+    if (Core::GetState() == Core::State::Uninitialized || Core::GetState() == Core::State::Paused)
+      g_controller_interface.UpdateInput();
+
+    if (Core::GetState() != Core::State::Stopping)
+    {
+      HotkeyManagerEmu::GetStatus();
+
+      if (!Core::IsRunningAndStarted())
+        continue;
+
+      // Fullscreen
+      if (IsHotkey(HK_FULLSCREEN))
+        m_parent->FullScreen();
```

Comment:
Like `MenuBar` and `ToolBar`, this should probably be a signal that `MainWindow` can connect to.

Ditto for the rest of the calls to `m_parent`, too.